dev/MRTCore/mrt/Core/src/MRM.cpp

```diff
@@ -899,26 +906,53 @@ STDAPI MrmGetFilePathFromName(_In_ PCWSTR filename, _Outptr_ PWSTR* filePath)
 
     pointerToPath = pathAllocated ? pathAllocated.get() : path;
 
-    size_t lengthInSizeT;
-    RETURN_IF_FAILED(StringCchLengthW(pointerToPath, STRSAFE_MAX_CCH, &lengthInSizeT));
-    length = static_cast<DWORD>(lengthInSizeT);
+    std::unique_ptr<wchar_t, decltype(&MrmFreeResource)> finalPath(nullptr, MrmFreeResource);
 
-    RETURN_IF_FAILED(StringCchLengthW(filename, STRSAFE_MAX_CCH, &lengthInSizeT));
-    DWORD lengthOfName = static_cast<DWORD>(lengthInSizeT);
+    // Will build the path at most twice.
+    // First time using current path. If not exist, do another time with parent path.
+    for (int i = 0; i < 2; i++)
+    {
+        size_t lengthInSizeT;
+        RETURN_IF_FAILED(StringCchLengthW(pointerToPath, STRSAFE_MAX_CCH, &lengthInSizeT));
+        length = static_cast<DWORD>(lengthInSizeT);
+
+        RETURN_IF_FAILED(StringCchLengthW(filename, STRSAFE_MAX_CCH, &lengthInSizeT));
+        DWORD lengthOfName = static_cast<DWORD>(lengthInSizeT);
+
+        RETURN_IF_FAILED(DWordAdd(length, lengthOfName, &length));
+        RETURN_IF_FAILED(DWordAdd(length, 1, &length));
+        RETURN_IF_FAILED(DWordMult(length, sizeof(wchar_t), &size));
+
+        PWSTR rawOutputPath = reinterpret_cast<PWSTR>(MrmAllocateBuffer(size));
+        RETURN_IF_NULL_ALLOC(rawOutputPath);
 
-    RETURN_IF_FAILED(DWordAdd(length, lengthOfName, &length));
-    RETURN_IF_FAILED(DWordAdd(length, 1, &length));
-    RETURN_IF_FAILED(DWordMult(length, sizeof(wchar_t), &size));
+        std::unique_ptr<wchar_t, decltype(&MrmFreeResource)> outputPath(rawOutputPath, MrmFreeResource);
+
+        RETURN_IF_FAILED(StringCchCopyW(outputPath.get(), length, pointerToPath));
+        RETURN_IF_FAILED(StringCchCatW(outputPath.get(), length, filename));
+
+        if (GetFileAttributes(outputPath.get()) != INVALID_FILE_ATTRIBUTES)
+        {
+            // The file exists. Done.
+            finalPath.swap(outputPath);
+            break;
+        }
 
-    PWSTR rawOutputPath = reinterpret_cast<PWSTR>(MrmAllocateBuffer(size));
-    RETURN_IF_NULL_ALLOC(rawOutputPath);
+        if (i == 0)
+        {
+            // If none of the file exists, will return the file in current path
+            finalPath.swap(outputPath);
+        }
 
-    std::unique_ptr<wchar_t, decltype(&MrmFreeResource)> outputPath(rawOutputPath, MrmFreeResource);
+        if (secondToLastSlash == nullptr)
+        {
+            break;
+        }
 
-    RETURN_IF_FAILED(StringCchCopyW(outputPath.get(), length, pointerToPath));
-    RETURN_IF_FAILED(StringCchCatW(outputPath.get(), length, filename));
+        *(secondToLastSlash + 1) = 0;
+    }
 
-    *filePath = outputPath.release();
+    *filePath = finalPath.release();
 
     return S_OK;
```

Comment:
Could we have the fallback impl be in a separate function that we call if filePath is empty after the existing code runs?

Reply:
not quite sure what you mean. but this has least code churn.

Reply:
The existing code looks in the current module path, right? We're adding looking in the parent folder too. I'm saying having the addition, i.e., looking in the parent folder, be in a separate function. That way if we want to remove it in the future, for example, it is much easier. It also makes the code more self-explanatory.

---
In reply to: [591992825](https://github.com/microsoft/ProjectReunion/pull/554#discussion_r591992825) [](ancestors = 591992825)

Reply:
then the second function will be similar to the first one. People will ask why you cannot abstract the logic. Plus, a few more places need to be updated. Similar logic of lookup/check existence/lookup again will also need to apply to multiple places. If you want to prepare for possible future removal, this is probably the easiest.

Reply:
ok, this is fine with me

---
In reply to: [591996471](https://github.com/microsoft/ProjectReunion/pull/554#discussion_r591996471) [](ancestors = 591996471)